zerver/lib/bugdown/__init__.py

```diff
@@ -1763,6 +1763,40 @@ def get_sub_registry(r: markdown.util.Registry, keys: List[str]) -> markdown.uti
 DEFAULT_BUGDOWN_KEY = -1
 ZEPHYR_MIRROR_BUGDOWN_KEY = -2
 
+class NormalizeWhitespace(markdown.preprocessors.NormalizeWhitespace):
+    # This class doesn't remove start and end of text characters from text.
+    # This is to allow HtmlStash placeholders to function properly on
+    # fenced codeblocks we find inside block level elements.
+
+    def run(self, lines: List[str]) -> List[str]:
+        source = '\n'.join(lines)
+        # source = source.replace(util.STX, "").replace(util.ETX, "")
+        source = source.replace("\r\n", "\n").replace("\r", "\n") + "\n\n"
+        source = source.expandtabs(self.md.tab_length)
+        source = re.sub(r'(?<=\n) +\n', '\n', source)
+        return source.split('\n')
+
+class BlockParser(markdown.blockprocessors.BlockParser):
```

Comment:
Similarly this needs a docstring explaining how it's modified from upstream -- is it to add a new feature, override a behavior, or something else?